### PR TITLE
Uncouple circular dependency with Codec2

### PR DESCRIPTION
**Pull request by tmiw**
Adds required files from Codec2 so that LPCNet doesn't have to be linked with Codec2 per https://github.com/drowe67/codec2/discussions/342. For those files, we compile with `-fvisibility=hidden` so that there aren't conflicts with the Codec2 versions when we link LPCNet into the former ([source](https://stackoverflow.com/questions/28588661/how-can-i-make-a-function-have-library-internal-linkage) for info about that compiler flag).

Current status: when linked with this PR, Codec2 ctests on master currently fail. However, it's unknown if these failures are related:

```
The following tests FAILED:
	 13 - test_FDMDV_modem_octave_mod_demod (Failed)
	 18 - test_COHPSK_modem_freq_offset (Failed)
	 21 - test_OFDM_modem_octave_port_Nc_31 (Failed)
	 74 - test_freedv_api_2020C_mpd (Failed)
```

Additionally, LPCNet ctests are failing due to use of developer specific paths (e.g. `build_linux` inside test_core_nn.sh):

```
40% tests passed, 3 tests failed out of 5

Total Test time (real) =   0.15 sec

The following tests FAILED:
	  1 - core_synthesis_default (Failed)
	  2 - core_synthesis_load_20h (Failed)
	  3 - core_synthesis_mag (Failed)
```

### CODE REVIEW

**Comment by tmiw** [comment]
Also, a lot more files needed to be brought in than I thought (though the majority are .h files).

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -19,8 +28,13 @@ set(lpcnet_freedv_srcs
...
 )
 
-add_library(lpcnetfreedv SHARED ${lpcnet_freedv_srcs})
-target_link_libraries(lpcnetfreedv codec2)
+# Set visibility of files brought in from Codec2 so that they're not visible to 
+# users (such as Codec2 itself).
+foreach(CODEC2_FILE IN LISTS codec2_import_srcs)
+    set_source_files_properties(${CODEC2_FILE} PROPERTIES COMPILE_FLAGS -fvisibility=hidden)
+endforeach()
+
```
Comment (drowe67) [inline src/CMakeLists.txt:36]: Gosh that's neat - I didn't know there was a way of looping in `cmake` :thinking: 

Well done on finding the magic `-fvisibility=hidden` linker option :+1:

**Comment by drowe67** [comment]
Thanks for working on this @tmiw.

These tests run OK for me (no path problems) but fail:
```
The following tests FAILED:
	  1 - core_synthesis_default (Failed)
	  2 - core_synthesis_load_20h (Failed)
	  3 - core_synthesis_mag (Failed)
```
I recall there were some numerical/precision issues which made some of these tests work on one machine and not another.  Not very good tests I guess.  Still, it would be nice to check we haven't broken anything :thinking:

**Comment by tmiw** [comment]
> Thanks for working on this @tmiw.
> 
> These tests run OK for me (no path problems) but fail:
> 
> ```
> The following tests FAILED:
> 	  1 - core_synthesis_default (Failed)
> 	  2 - core_synthesis_load_20h (Failed)
> 	  3 - core_synthesis_mag (Failed)
> ```
> 
> I recall there were some numerical/precision issues which made some of these tests work on one machine and not another. Not very good tests I guess. Still, it would be nice to check we haven't broken anything 🤔

Well, at least the paths seem to be more platform/developer independent now, but still no dice on my end. 

FWIW, the test that's currently being done in GitHub actions/freedv-gui builds is this:

```
# simple test to make sure the code runs
cd src && sox ../../wav/wia.wav -t raw -r 16000 - | ./lpcnet_enc -s | ./lpcnet_dec -s > /dev/null
```

**Comment by drowe67** [comment]
I manually checked the feature file is the same for `master` and `ms-uncouple-codec2`:
```
~/tmp/LPCNet/build_linux ms-uncouple-codec2 $ ./src/dump_data --test --c2pitch ../wav/birch.wav birch.f32
~/tmp/LPCNet/build_linux ms-uncouple-codec2 $ diff birch.f32 ~/LPCNet/unittest/birch.f32
```
The pitch estimate is in one of the fields in the `.f32` file, so this suggests this PR is doing the same thing as `master`, which uses the `libcodec2` pitch estimation code :+1:

File: src/codec2_pitch.c [hunk 2]
```diff
@@ -59,19 +32,6 @@ struct CODEC2_PITCH_S {
     COMP          W[FFT_ENC];	        /* DFT of w[] */
 };
```
Comment (drowe67) [inline src/codec2_pitch.c:33]: I noticed some warnings when building - think `W` should be a float, e.g. 
``` 
void make_analysis_window(C2CONST *c2const, codec2_fft_cfg fft_fwd_cfg, float w[], float W[]);
```

Also noticed some other unused variable warnings - but we should probably work on these in the codec2 repo.

**Comment by drowe67** [comment]
Should the LPCNet `README.md` build instructions be tweaked?

**Comment by tmiw** [comment]
> Should the LPCNet `README.md` build instructions be tweaked?

Yep, just did. 👍

File: src/codec2_pitch.c (hunk 2 above)
Comment (tmiw) [reply src/codec2_pitch.c:33]: OK, codec2_pitch.c warnings have been resolved. There are a few more in LPCNet itself but that could be a separate PR if desired:

```
[  6%] Building C object src/CMakeFiles/lpcnetfreedv.dir/pitch.c.o
/Users/mooneer/devel/LPCNet/src/pitch.c:229:37: warning: performing pointer subtraction with a null pointer has undefined behavior [-Wnull-pointer-subtraction]
   celt_assert((((unsigned char *)_x-(unsigned char *)NULL)&3)==0);
   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/mooneer/devel/LPCNet/src/arch.h:63:34: note: expanded from macro 'celt_assert'
#define celt_assert(cond) {if (!(cond)) {celt_fatal("assertion failed: " #cond);}}
                                 ^~~~
1 warning generated.
...
[ 12%] Building C object src/CMakeFiles/lpcnetfreedv.dir/lpcnet_dump.c.o
/Users/mooneer/devel/LPCNet/src/lpcnet_dump.c:120:9: warning: variable 'E' set but not used [-Wunused-but-set-variable]
  float E = 0;
        ^
1 warning generated.
...
[ 35%] Building C object src/CMakeFiles/dump_data.dir/dump_data.c.o
/Users/mooneer/devel/LPCNet/src/dump_data.c:115:9: warning: variable 'E' set but not used [-Wunused-but-set-variable]
  float E = 0;
        ^
1 warning generated.
```

File: unittest/test_core_nn.sh.in [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+# test generation of training data (doesn't really test training as that takes hours)
+# TODO: This test not working yet
+if [ ! -z $TRAIN_TEST ]; then
+    TRAIN_SRC=all_speech
+    TRAIN_OUT_PCM=$(mktemp).pcm
+    TRAIN_OUT_F32=$(mktemp).f32
+    TRAIN_TARG_PCM=all_speech.pcm
+    TRAIN_TARG_F32=all_speech.f32
+    @CMAKE_CURRENT_BINARY_DIR@/src/dump_data --train --c2pitch -z 0 -n 1E6 ~/Downloads/$TRAIN_SRC.sw $TRAIN_OUT_F32 $TRAIN_OUT_PCM
```
Comment (drowe67) [inline unittest/test_core_nn.sh.in]: I'm not fond of the `.in` -> `.sh` model as it means the test can't be run manually without running `cmake` after every edit.  There are some more convenient models in the `codec2` tests, e.g. setting the `$PATH` externally or passing in a build time path variable.

Having said that - most of these tests don't work anyway.  I'll take an action to add a simple test that at least tests the changes in this PR.

**Comment by drowe67** [comment]
@tmiw In https://github.com/drowe67/LPCNet/pull/44 I've had a go at building up some ctests that pass and can run with GitHub actions.  Please try merging this PR with `LPCNet/master` and lets see if the tests work with this PR.  The first ctest should pick up any pitch estimator issues, as that's one of the features `dump_data` generates.

**Comment by tmiw** [comment]
> @tmiw In #44 I've had a go at building up some ctests that pass and can run with GitHub actions. Please try merging this PR with `LPCNet/master` and lets see if the tests work with this PR. The first ctest should pick up any pitch estimator issues, as that's one of the features `dump_data` generates.

All tests pass for me now. 👍

**Comment by drowe67** [comment]
@tmiw OK to merge now?  I bumped the version number, as I figure this will need a new release.

**Comment by tmiw** [comment]
> @tmiw OK to merge now? I bumped the version number, as I figure this will need a new release.

Yep.

**Comment by tmiw** [comment]
Should a new release be tagged at https://github.com/drowe67/LPCNet/releases?

**Comment by drowe67** [comment]
> Should a new release be tagged at https://github.com/drowe67/LPCNet/releases?

Yes, but how about we wait until `codec2` and `freedv-gui` have been updated in line with this change?  There might still be some `LPCNet` tweaks required before the dust settles.

**Comment by tmiw** [comment]
> > Should a new release be tagged at https://github.com/drowe67/LPCNet/releases?
> 
> Yes, but how about we wait until `codec2` and `freedv-gui` have been updated in line with this change? There might still be some `LPCNet` tweaks required before the dust settles.

That works. I think there's an issue compiling freedv-gui for Windows so I'll investigate that.

**Pull request merged**